File: app/src/main/java/com/productions/gizzmoo/pokemonpuzzleleague/TrainerPreference.java
package com.productions.gizzmoo.pokemonpuzzleleague;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Parcelable;
import android.preference.DialogPreference;
import android.util.AttributeSet;
import android.view.View;
import android.widget.AdapterView;
import android.widget.GridView;
import android.widget.Toast;

/**
 * Created by Chrystian on 4/8/2018.
 */

public class TrainerPreference extends DialogPreference {
    private static final int DEFAULT_ID = 0; // Ash
    private int mTrainerID = DEFAULT_ID;

    private GridView mGridView;
    private Context mContext;
    private ImageAdapter mAdapter;

    private Bitmap[] mBitmaps;
    private String[] mTrainerNames;

    public TrainerPreference(Context context) {
        this(context, null, 0);
    }

    public TrainerPreference(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public TrainerPreference(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        mContext = context;
        setDialogLayoutResource(R.layout.trainer_preference);
        init();
    }


    @Override
    protected void onBindDialogView(View view) {
        mGridView = view.findViewById(R.id.grid);
        mAdapter = new ImageAdapter(mContext, mBitmaps);
        mAdapter.positionChosen(mTrainerID);
        mGridView.setAdapter(mAdapter);
        mGridView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            public void onItemClick(AdapterView<?> parent, View v, int position, long id) {
                persistInt(position);
                mTrainerID = position;
                Toast.makeText(mContext, "" + mTrainerNames[position], Toast.LENGTH_SHORT).show();
                getDialog().dismiss();
            }
        });

        super.onBindDialogView(view);
    }

    @Override
    protected void onSetInitialValue(boolean restorePersistedValue, Object defaultValue) {
        if (restorePersistedValue) {
            mTrainerID = getPersistedInt(DEFAULT_ID);
        } else {
            persistInt(mTrainerID);
        }
    }

    @Override
    protected void onDialogClosed(boolean positiveResult) {
        if (positiveResult) {
            persistInt(mTrainerID);
        }
    }

    @Override
    protected Parcelable onSaveInstanceState() {
        final Parcelable superState = super.onSaveInstanceState();
        if (isPersistent()) {
            return superState;
        }

        final SavedState myState = new SavedState(superState);
        myState.value = mTrainerID;
        return myState;
    }

    @Override
    protected void onRestoreInstanceState(Parcelable state) {
        if (state == null || !state.getClass().equals(SavedState.class)) {
            super.onRestoreInstanceState(state);
            return;
        }

        SavedState myState = (SavedState) state;
        super.onRestoreInstanceState(myState.getSuperState());

        mTrainerID = myState.value;
    }

    private void init() {
        int[] imageResources = TrainerResources.Companion.getAllTrainerPortraits();
        mBitmaps = new Bitmap[imageResources.length];

        for (int i = 0; i < imageResources.length; i++) {
            mBitmaps[i] = BitmapFactory.decodeResource(mContext.getResources(), imageResources[i]);
        }

        mTrainerNames = TrainerResources.Companion.getTrainerNames(mContext);
    }
}

File: app/src/main/java/com/productions/gizzmoo/pokemonpuzzleleague/settings/TrainerPreference.kt
package com.productions.gizzmoo.pokemonpuzzleleague.settings

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.os.Parcelable
import androidx.preference.DialogPreference
import android.util.AttributeSet
import com.productions.gizzmoo.pokemonpuzzleleague.TrainerResources

class TrainerPreference(context: Context, attrs: AttributeSet? = null) : DialogPreference(context, attrs) {
    val bitmaps: Array<Bitmap>
    var trainerID = DEFAULT_ID
        private set
    private var trainerNames: Array<String>

    init {
        val imageResources = TrainerResources.getAllTrainerPortraits()
        bitmaps = Array(imageResources.size) { i -> BitmapFactory.decodeResource(context.resources, imageResources[i])}
        trainerNames = TrainerResources.getTrainerNames(context)
    }

    override fun onSetInitialValue(restorePersistedValue: Boolean, defaultValue: Any?) {
        if (restorePersistedValue) {
            trainerID = getPersistedInt(DEFAULT_ID)
        } else {
            persistInt(trainerID)
        }
    }

    override fun onSaveInstanceState(): Parcelable {
        val superState = super.onSaveInstanceState()
        val myState = SavedState(superState)
        myState.value = trainerID
        return myState
    }

    override fun onRestoreInstanceState(state: Parcelable?) {
        if (state == null || state !is SavedState) {
            super.onRestoreInstanceState(state)
            return
        }

        val myState = state as SavedState?
        super.onRestoreInstanceState(myState?.superState)

        trainerID = myState?.value ?: DEFAULT_ID
    }

    override fun getSummary(): CharSequence {
        return trainerNames[trainerID]
    }

    fun persistTrainerID(newTrainerID: Int) {
        trainerID = newTrainerID
        persistInt(newTrainerID)
        summary = trainerNames[trainerID]
    }

    companion object {
        const val DEFAULT_ID = 0 // Ash
    }
}